File: source/programing_exercises/ruby/fizz_buzz.rb
# frozen_string_literal: true

class FizzBuzz
  def self.generate(number)
    value = number

    if (value % 3 == 0) && (value % 5 == 0)
      value = 'FizzBuzz'
    elsif value % 3 == 0
      value = 'Fizz'
    elsif value % 5 == 0
      value = 'Buzz'
    end

    value
  end

  def self.iterate(count)
    array = []
    count.times do |n|
      n += 1
      array.push(FizzBuzz.generate(n))
    end
    array
  end
end

require 'minitest/autorun'
class FizzBuzzSpec < Minitest::Spec
  describe FizzBuzz do
    it '3ならばFizzを返す' do
      expect(FizzBuzz.generate(3)).must_equal 'Fizz'
    end

    it '9ならばFizzを返す' do
      expect(FizzBuzz.generate(9)).must_equal 'Fizz'
    end

    it '5ならばBuzzを返す' do
      expect(FizzBuzz.generate(5)).must_equal 'Buzz'
    end

    it '10ならばBuzzを返す' do
      expect(FizzBuzz.generate(10)).must_equal 'Buzz'
    end

    it '50ならばBuzzを返す' do
      expect(FizzBuzz.generate(50)).must_equal 'Buzz'
    end

    it '15ならばFizzBuzzを返す' do
      expect(FizzBuzz.generate(15)).must_equal 'FizzBuzz'
    end

    it '30ならばFizzBuzzを返す' do
      expect(FizzBuzz.generate(30)).must_equal 'FizzBuzz'
    end

    it '5回実行されたならば配列を返す' do
      expect(FizzBuzz.iterate(5)).must_equal [1, 2, 'Fizz', 4, 'Buzz']
    end

    it '10回実行されたならば配列を返す' do
      expect(FizzBuzz.iterate(10)).must_equal [1, 2, 'Fizz', 4, 'Buzz', 'Fizz', 7, 8, 'Fizz', 'Buzz']
    end
  end
end

File: source/programing_exercises/java/src/main/java/FizzBuzz.java
interface IType{
    String generate(int number);
}

class Type01 implements IType {
    @Override
    public String generate(int number) {
        return Integer.toString(number);
    }
}

class Type02 implements IType {
    @Override
    public String generate(int number) {
        return "Fizz";
    }
}

class Type03 implements IType {
    @Override
    public String generate(int number) {
        return "Buzz";
    }
}

class Type04 implements IType {
    @Override
    public String generate(int number) {
        if (number % 3 == 0 && number % 5 == 0) {
            return "FIZZBUZZ";
        } else if (number % 3 == 0) {
            return "FIZZ";
        } else if (number % 5 == 0) {
            return "BUZZ";
        }
        return Integer.toString(number);
    }
}

class Type05 implements IType {
    @Override
    public String generate(int number) {
        if (number % 2 == 0 && number % 3 == 0) {
            return "FIZZBUZZ";
        } else if (number % 2 == 0) {
            return "Fizz";
        } else if (number % 3 == 0) {
            return "Buzz";
        }
        return Integer.toString(number);
    }
}

class TypeNormal implements IType {
    @Override
    public String generate(int number) {
        if (number % 3 == 0 && number % 5 == 0) {
            return "FizzBuzz";
        } else if (number % 3 == 0) {
            return "Fizz";
        } else if (number % 5 == 0) {
            return "Buzz";
        }
        return Integer.toString(number);
    }
}

enum FizzBuzzType {
    one(new Type01()),
    two(new Type02()),
    three(new Type03()),
    four(new Type04()),
    five(new Type05()),
    normal(new TypeNormal());

    private IType type;

    private FizzBuzzType(IType type ) {
        this.type = type;
    }

    String generate(int number) {
        return type.generate(number);
    }    
}

class FizzBuzz {
    private String _value;
    private String _values[];

    public String getValue() {
        return this._value;
    }

    public String[] getValues() {
        return this._values;
    }

    public String[] iterate(int count) {
        this._values = new String[count];
        FizzBuzzType fizzBuzz = FizzBuzzType.valueOf("normal");

        for (int i = 0; i < count; i = i + 1) {            
            this._values[i] = fizzBuzz.generate(i + 1);
        }

        return this._values;
    }
}
